src/commands/subcommand.cc

```diff
@@ -476,65 +476,23 @@ std::ostream & SubcommandOptions::showHelpOn( std::ostream & out ) const
   return out;	// FAKE!
 }
 
-///////////////////////////////////////////////////////////////////
-namespace
+SubCmd::SubCmd(std::vector<std::string> &&commandAliases_r , boost::shared_ptr<SubcommandOptions> options_r) :
+  ZypperBaseCommand (
+    std::move( commandAliases_r ),
+    "subcommand",
+    // translators: command summary: subcommand
+    _("Lists available subcommands."),
+    "", //no help text, its created on demand
+    DisableAll
+    ),
+  _options ( options_r )
 {
-  ///////////////////////////////////////////////////////////////////
-  /// \class SubcommandImpl
-  /// \brief Implementation of subcommands.
-  ///////////////////////////////////////////////////////////////////
-  class SubcommandImpl : public CommandBase<SubcommandImpl,SubcommandOptions>
-  {
-    typedef CommandBase<SubcommandImpl,SubcommandOptions> CommandBase;
-  public:
-    /** Standard ctor for subcommand using zypper_r.commandOptionsAs. */
-    SubcommandImpl( Zypper & zypper_r ) : CommandBase( zypper_r ) {}
-    /** \overload using custom options_r. */
-    SubcommandImpl( Zypper & zypper_r, shared_ptr<SubcommandOptions> options_r ) : CommandBase( zypper_r, options_r ) {}
-    // CommandBase::_zypper
-    // CommandBase::options;	// access/manip command options
-    // CommandBase::run;	// action + catch and repost Out::Error
-    // CommandBase::execute;	// run + final "Done"/"Finished with error" message
-    // CommandBase::showHelp;// Show user help on command
-  public:
-    /** default action */
-    void action();
-
-  private:
-    /** Store command in \c _args[0]. */
-    void setArg0( std::string arg0_r )
-    {
-      if ( options()._args.empty() )
-	options()._args.push_back( std::move(arg0_r) );
-      else
-	options()._args[0] = std::move(arg0_r);
-    }
-  };
-  ///////////////////////////////////////////////////////////////////
-
-  void SubcommandImpl::action()
-  {
-    setArg0( (options()._detected._path / options()._detected._name).asString() );
-    RunCommand cmd( options()._args );
-    if ( cmd.run() != 0 )
-      throw( Out::Error( cmd.exitStatus(), cmd.execError() ) );
+  if ( !_options ) {
+    _options.reset ( new SubcommandOptions() );
   }
-} // namespace
-///////////////////////////////////////////////////////////////////
-
-int subcommand( Zypper & zypper_r )
-{
-  zypper_r.cleanupForSubcommand();
-  return SubcommandImpl( zypper_r ).run();
-}
-
-int subcommand( Zypper & zypper_r, shared_ptr<SubcommandOptions> options_r )
-{
-  zypper_r.cleanupForSubcommand();
```

Comment:
Looks like the `zypper_r.cleanupForSubcommand();` call is lost (here and from L527). 
Should probably be added as 1st action to `SubCmd::runCmd`.